pycqed/scripts/Spectroscopy.py

```diff
@@ -0,0 +1,281 @@
+                                             extract_only=extract_only)
+        self.extract_fitparams = self.options_dict.get('fitparams',True)
+        self.params_dict = {'freq':'sweep_points',
+                            'amp':'amp',
+                            'phase':'phase'}
+        if self.extract_fitparams:
+            self.params_dict.update({'fitparams':'fit_params'})
+
+        self.numeric_params = ['freq', 'amp', 'phase']
+        if 'qubit_label' in self.options_dict:
+            self.labels.extend(self.options_dict['qubit_label'])
+        if auto is True:
+            self.run_analysis()
+
+    def process_data(self):
+        self.freq_label = 'Frequency (GHz)'
+        self.amp_label = 'Transmission amplitude (arb. units)'
+
+        self.phase_label = 'Transmission phase (degrees)'
+        self.freq_range = self.options_dict.get('freq_range',None)
+        self.amp_range = self.options_dict.get('amp_range',None)
+        self.phase_range = self.options_dict.get('phase_range',None)
+        self.plotsize = self.options_dict.get('plotsize',(8,5))
+        if len(self.data_dict['timestamps']) == 1:
+            self.plot_frequency = np.squeeze(self.data_dict['freq'])
+            self.plot_amp = np.squeeze(self.data_dict['amp'])
+            self.plot_phase = np.squeeze(self.data_dict['phase'])
+            print(self.plot_frequency.shape, self.plot_amp.shape, self.plot_phase.shape)
+
+        else:
+            #TRANSPOSE ALSO NEEDS TO BE CODED FOR 2D
+            self.plot_xvals = self.options_dict.get('xvals',np.array([[tt] for tt in range(len(self.data_dict['timestamps']))]))
+            self.plot_xwidth = self.options_dict.get('xwidth',None)
+            self.plot_xlabel = self.options_dict.get('xlabel','Scan number')
+            if self.plot_xwidth == 'auto':
+                x_diff = np.diff(np.ravel(self.plot_xvals))
+                dx1 = np.concatenate(([x_diff[0]],x_diff))
+                dx2 = np.concatenate((x_diff,[x_diff[-1]]))
+                self.plot_xwidth = np.maximum(dx1,dx2)
+                self.plot_frequency = np.array([[tt] for tt in self.data_dict['freq']])
+                self.plot_phase = np.array([[tt] for tt in self.data_dict['phase']])
+                self.plot_amp = np.array([np.array([tt]).transpose() for tt in self.data_dict['amp']])
+
+            else:
+                self.plot_frequency = self.data_dict['freq']
+                self.plot_phase = self.data_dict['phase']
+                self.plot_amp = self.data_dict['amp']
+
+
+
+
+
+
+
+    def prepare_plots(self):
+
+        if len(self.data_dict['timestamps']) == 1:
+            plot_fn = self.plot_line
+            self.plot_dicts['amp'] = {'plotfn': plot_fn,
+                            'xvals': self.plot_frequency,
+                            'yvals': self.plot_amp,
+                            'title': 'Spectroscopy amplitude: %s'%(self.timestamps[0]),
```

Comment:
I suggest we avoid any references to `self."some attribute"` if it can be avoided. In this case the timestamps also exist in the data_dict. 

As a matter of fact (I see more of these above) I would avoid using attribute assignment in the self. completely and stuck exclusively to the data_dict.